In [32]:
import pandas as pd

file_path = "../data/raw/premier_league_complete_stats_until31thGameDayOnSeason2025-26.csv"

df = pd.read_csv(file_path, encoding="latin-1")

df



,player_name,team_name,position,accurateCrosses,accurateCrossesPercentage,accurateFinalThirdPasses,accurateLongBalls,accurateLongBallsPercentage,accurateOppositionHalfPasses,accurateOwnHalfPasses,...,totalDuelsWonPercentage,totalLongBalls,totalOppositionHalfPasses,totalOwnHalfPasses,totalPasses,totalShots,totwAppearances,touches,wasFouled,yellowCards
0,Declan Rice,Arsenal,M,41,28.082192,459,72,48.648649,891,687,...,61.111111,148,1054,743,1797,36,3,2327,8,2
1,Gabriel Magalhães,Arsenal,D,1,20.000000,180,63,46.323529,512,747,...,70.334928,136,622,797,1419,18,5,1747,27,3
2,Bukayo Saka,Arsenal,F,36,32.432432,300,14,36.842105,419,105,...,53.020134,38,557,126,683,63,4,1236,51,1
3,Martín Zubimendi,Arsenal,M,2,18.181818,369,48,49.484536,833,661,...,60.189573,97,985,709,1694,26,2,2026,10,4
4,Max Dowman,Arsenal,M,0,0.000000,9,0,0.000000,13,3,...,42.857143,0,14,3,17,5,0,51,4,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
512,Tolu Arokodare,Wolverhampton,F,0,0.000000,50,1,16.666667,102,49,...,46.825397,6,169,73,242,39,0,529,13,1
513,Fer López,Wolverhampton,M,1,25.000000,31,1,25.000000,58,17,...,31.250000,4,80,23,103,5,0,156,4,1
514,Matt Doherty,Wolverhampton,D,2,28.571429,38,12,38.709677,91,125,...,50.666667,31,125,143,268,3,0,386,5,4
515,Toti Gomes,Wolverhampton,D,3,30.000000,52,23,35.384615,175,357,...,60.810811,65,236,381,617,4,0,786,8,2


In [33]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

Rows: 517
Columns: 99


In [34]:
df.columns.tolist()

['player_name',
 'team_name',
 'position',
 'accurateCrosses',
 'accurateCrossesPercentage',
 'accurateFinalThirdPasses',
 'accurateLongBalls',
 'accurateLongBallsPercentage',
 'accurateOppositionHalfPasses',
 'accurateOwnHalfPasses',
 'accuratePasses',
 'accuratePassesPercentage',
 'aerialDuelsWon',
 'aerialDuelsWonPercentage',
 'aerialLost',
 'appearances',
 'assists',
 'ballRecovery',
 'bigChancesCreated',
 'bigChancesMissed',
 'blockedShots',
 'cleanSheet',
 'clearances',
 'dispossessed',
 'dribbledPast',
 'duelLost',
 'errorLeadToGoal',
 'errorLeadToShot',
 'expectedAssists',
 'expectedGoals',
 'fouls',
 'freeKickGoal',
 'goalConversionPercentage',
 'goals',
 'goalsAssistsSum',
 'goalsConceded',
 'goalsConcededInsideTheBox',
 'goalsConcededOutsideTheBox',
 'goalsFromInsideTheBox',
 'goalsFromOutsideTheBox',
 'goalsPrevented',
 'groundDuelsWon',
 'groundDuelsWonPercentage',
 'headedGoals',
 'highClaims',
 'hitWoodwork',
 'inaccuratePasses',
 'interceptions',
 'keyPasses',
 'leftFoo

In [35]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 517 entries, 0 to 516
Data columns (total 99 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   player_name                   517 non-null    str    
 1   team_name                     517 non-null    str    
 2   position                      517 non-null    str    
 3   accurateCrosses               517 non-null    int64  
 4   accurateCrossesPercentage     517 non-null    float64
 5   accurateFinalThirdPasses      517 non-null    int64  
 6   accurateLongBalls             517 non-null    int64  
 7   accurateLongBallsPercentage   517 non-null    float64
 8   accurateOppositionHalfPasses  517 non-null    int64  
 9   accurateOwnHalfPasses         517 non-null    int64  
 10  accuratePasses                517 non-null    int64  
 11  accuratePassesPercentage      517 non-null    float64
 12  aerialDuelsWon                517 non-null    int64  
 13  aerialDuelsWonPe

In [36]:
df["position"].value_counts()

position
M    213
D    172
F     97
G     35
Name: count, dtype: int64

In [37]:
[c for c in df.columns if "minute" in c.lower()]

['minutesPlayed']

In [38]:
df["minutesPlayed"].describe()

count     517.000000
mean     1188.214700
std       844.914132
min         1.000000
25%       412.000000
50%      1132.000000
75%      1896.000000
max      2790.000000
Name: minutesPlayed, dtype: float64

In [39]:
(df["minutesPlayed"] >= 900).sum()

np.int64(301)

In [40]:
df[df["minutesPlayed"] >= 900]["position"].value_counts()

position
M    123
D    108
F     47
G     23
Name: count, dtype: int64

In [41]:
df["player_name"].duplicated().sum()

np.int64(0)

In [42]:
df.duplicated().sum()

np.int64(0)

## Player Eligibility

For the performance analysis, players are included only if they have played at least 900 minutes.

This threshold is used to reduce the influence of very small playing-time samples while retaining a sufficiently large player pool for positional analysis.

- Total players in raw dataset: 517
- Eligible players: 301
- Minimum playing time: 900 minutes

In [43]:
eligible_df = df[df["minutesPlayed"] >= 900].copy()

In [44]:
eligible_df.shape

(301, 99)

In [45]:
eligible_df.select_dtypes(include="number").columns.tolist()

['accurateCrosses',
 'accurateCrossesPercentage',
 'accurateFinalThirdPasses',
 'accurateLongBalls',
 'accurateLongBallsPercentage',
 'accurateOppositionHalfPasses',
 'accurateOwnHalfPasses',
 'accuratePasses',
 'accuratePassesPercentage',
 'aerialDuelsWon',
 'aerialDuelsWonPercentage',
 'aerialLost',
 'appearances',
 'assists',
 'ballRecovery',
 'bigChancesCreated',
 'bigChancesMissed',
 'blockedShots',
 'cleanSheet',
 'clearances',
 'dispossessed',
 'dribbledPast',
 'duelLost',
 'errorLeadToGoal',
 'errorLeadToShot',
 'expectedAssists',
 'expectedGoals',
 'fouls',
 'freeKickGoal',
 'goalConversionPercentage',
 'goals',
 'goalsAssistsSum',
 'goalsConceded',
 'goalsConcededInsideTheBox',
 'goalsConcededOutsideTheBox',
 'goalsFromInsideTheBox',
 'goalsFromOutsideTheBox',
 'goalsPrevented',
 'groundDuelsWon',
 'groundDuelsWonPercentage',
 'headedGoals',
 'highClaims',
 'hitWoodwork',
 'inaccuratePasses',
 'interceptions',
 'keyPasses',
 'leftFootGoals',
 'matchesStarted',
 'minutesPlayed

In [46]:
for i, col in enumerate(eligible_df.columns):
    print(i, col)

0 player_name
1 team_name
2 position
3 accurateCrosses
4 accurateCrossesPercentage
5 accurateFinalThirdPasses
6 accurateLongBalls
7 accurateLongBallsPercentage
8 accurateOppositionHalfPasses
9 accurateOwnHalfPasses
10 accuratePasses
11 accuratePassesPercentage
12 aerialDuelsWon
13 aerialDuelsWonPercentage
14 aerialLost
15 appearances
16 assists
17 ballRecovery
18 bigChancesCreated
19 bigChancesMissed
20 blockedShots
21 cleanSheet
22 clearances
23 dispossessed
24 dribbledPast
25 duelLost
26 errorLeadToGoal
27 errorLeadToShot
28 expectedAssists
29 expectedGoals
30 fouls
31 freeKickGoal
32 goalConversionPercentage
33 goals
34 goalsAssistsSum
35 goalsConceded
36 goalsConcededInsideTheBox
37 goalsConcededOutsideTheBox
38 goalsFromInsideTheBox
39 goalsFromOutsideTheBox
40 goalsPrevented
41 groundDuelsWon
42 groundDuelsWonPercentage
43 headedGoals
44 highClaims
45 hitWoodwork
46 inaccuratePasses
47 interceptions
48 keyPasses
49 leftFootGoals
50 matchesStarted
51 minutesPlayed
52 offsides
53 o

In [47]:
eligible_df[
    [
        "player_name",
        "position",
        "minutesPlayed",
        "goals",
        "assists",
        "totalShots",
        "shotsOnTarget",
        "accuratePasses",
        "accuratePassesPercentage"
    ]
].head(10)

,player_name,position,minutesPlayed,goals,assists,totalShots,shotsOnTarget,accuratePasses,accuratePassesPercentage
0,Declan Rice,M,2585,4,5,36,12,1578,87.813022
1,Gabriel Magalhães,D,2166,3,4,18,5,1259,88.724454
2,Bukayo Saka,F,2001,6,3,63,26,524,76.720351
3,Martín Zubimendi,M,2604,5,1,26,9,1494,88.193625
5,Martin Ødegaard,M,1078,1,5,20,7,537,85.103011
6,Eberechi Eze,F,1504,6,2,49,16,437,84.200385
7,William Saliba,D,2075,1,0,5,1,1623,92.689891
8,David Raya,G,2790,0,0,0,0,604,66.519824
9,Piero Hincapié,D,1504,1,2,2,1,662,89.218329
11,Jurriën Timber,D,2457,3,5,25,7,925,85.253456


In [48]:
eligible_df["goals_per90"] = (
    eligible_df["goals"] / eligible_df["minutesPlayed"]
) * 90

In [49]:
eligible_df[["player_name", "minutesPlayed", "goals", "goals_per90"]].head(10)

,player_name,minutesPlayed,goals,goals_per90
0,Declan Rice,2585,4,0.139265
1,Gabriel Magalhães,2166,3,0.124654
2,Bukayo Saka,2001,6,0.269865
3,Martín Zubimendi,2604,5,0.172811
5,Martin Ødegaard,1078,1,0.083488
6,Eberechi Eze,1504,6,0.359043
7,William Saliba,2075,1,0.043373
8,David Raya,2790,0,0.000000
9,Piero Hincapié,1504,1,0.059840
11,Jurriën Timber,2457,3,0.109890


In [50]:
eligible_df["assists_per90"] = (
    eligible_df["assists"] / eligible_df["minutesPlayed"]
) * 90

In [51]:
eligible_df[
    [
        "player_name",
        "position",
        "goals_per90",
        "assists_per90"
    ]
].head(10)

,player_name,position,goals_per90,assists_per90
0,Declan Rice,M,0.139265,0.174081
1,Gabriel Magalhães,D,0.124654,0.166205
2,Bukayo Saka,F,0.269865,0.134933
3,Martín Zubimendi,M,0.172811,0.034562
5,Martin Ødegaard,M,0.083488,0.417440
6,Eberechi Eze,F,0.359043,0.119681
7,William Saliba,D,0.043373,0.000000
8,David Raya,G,0.000000,0.000000
9,Piero Hincapié,D,0.059840,0.119681
11,Jurriën Timber,D,0.109890,0.183150


In [52]:
eligible_df.groupby("position")[["goals", "assists", "totalShots", "accuratePasses"]].mean()

,goals,assists,totalShots,accuratePasses
position,,,,
D,0.879630,1.064815,12.546296,835.138889
F,6.063830,1.574468,41.127660,272.297872
G,0.000000,0.086957,0.043478,533.304348
M,2.796748,2.365854,26.682927,670.252033


In [53]:
[c for c in eligible_df.columns if any(
    word in c.lower()
    for word in ["tackle", "interception", "clearance", "block", "duel"]
)]

['aerialDuelsWon',
 'aerialDuelsWonPercentage',
 'blockedShots',
 'clearances',
 'duelLost',
 'groundDuelsWon',
 'groundDuelsWonPercentage',
 'interceptions',
 'tackles',
 'tacklesWon',
 'tacklesWonPercentage',
 'totalDuelsWon',
 'totalDuelsWonPercentage']

In [54]:
eligible_df[
    [
        "tackles",
        "tacklesWon",
        "tacklesWonPercentage",
        "interceptions",
        "clearances",
        "blockedShots",
        "aerialDuelsWon",
        "aerialDuelsWonPercentage",
        "groundDuelsWon",
        "groundDuelsWonPercentage"
    ]
].describe().T

,count,mean,std,min,25%,50%,75%,max
tackles,301.0,30.049834,19.183174,0.0,16.000000,28.000000,40.000000,96.0
tacklesWon,301.0,17.760797,11.777489,0.0,10.000000,16.000000,23.000000,56.0
tacklesWonPercentage,301.0,57.059995,17.105447,0.0,51.162791,58.461538,65.517241,100.0
interceptions,301.0,15.079734,10.779315,0.0,7.000000,13.000000,23.000000,52.0
clearances,301.0,50.491694,45.897757,0.0,19.000000,35.000000,67.000000,233.0
blockedShots,301.0,6.465116,5.640001,0.0,3.000000,5.000000,9.000000,27.0
aerialDuelsWon,301.0,29.750831,26.328585,0.0,11.000000,21.000000,42.000000,147.0
aerialDuelsWonPercentage,301.0,49.603035,17.953097,0.0,37.500000,49.295775,59.090909,100.0
groundDuelsWon,301.0,60.036545,33.529619,0.0,37.000000,57.000000,79.000000,195.0
groundDuelsWonPercentage,301.0,53.190854,13.417944,0.0,45.000000,51.937984,58.878505,100.0


In [55]:
eligible_df[
    [
        "player_name",
        "position",
        "tackles",
        "tacklesWonPercentage",
        "interceptions",
        "clearances",
        "blockedShots",
        "aerialDuelsWon",
        "aerialDuelsWonPercentage"
    ]
].query("position == 'D'").head(15)

,player_name,position,tackles,tacklesWonPercentage,interceptions,clearances,blockedShots,aerialDuelsWon,aerialDuelsWonPercentage
1,Gabriel Magalhães,D,32,50.000000,18,149,6,87,66.923077
7,William Saliba,D,29,58.620690,14,102,1,60,54.054054
9,Piero Hincapié,D,39,53.846154,20,75,1,30,51.724138
11,Jurriën Timber,D,66,56.060606,24,53,3,40,66.666667
16,Riccardo Calafiori,D,31,58.064516,14,48,9,40,58.823529
27,Marc Guéhi,D,37,64.864865,39,137,7,82,68.907563
30,Rúben Dias,D,13,46.153846,22,115,3,58,65.168539
34,Joško Gvardiol,D,18,72.222222,13,60,4,26,56.521739
38,Matheus Nunes,D,48,70.833333,25,82,1,30,56.603774
42,Nico O'Reilly,D,61,52.459016,17,56,9,38,54.285714


In [56]:
eligible_df[
    [
        "player_name",
        "position",
        "groundDuelsWon",
        "groundDuelsWonPercentage",
        "totalDuelsWon",
        "totalDuelsWonPercentage"
    ]
].query("position != 'G'").head(15)

,player_name,position,groundDuelsWon,groundDuelsWonPercentage,totalDuelsWon,totalDuelsWonPercentage
0,Declan Rice,M,77,56.204380,121,61.111111
1,Gabriel Magalhães,D,60,75.949367,147,70.334928
2,Bukayo Saka,F,136,54.618474,158,53.020134
3,Martín Zubimendi,M,68,54.400000,127,60.189573
5,Martin Ødegaard,M,33,46.478873,35,44.871795
6,Eberechi Eze,F,69,57.500000,77,51.333333
7,William Saliba,D,48,60.000000,108,56.544503
9,Piero Hincapié,D,55,61.797753,85,57.823129
11,Jurriën Timber,D,107,54.314721,147,57.198444
12,Leandro Trossard,F,72,44.171779,78,43.820225


In [57]:
eligible_df["groundDuelsWon_per90"] = (
    eligible_df["groundDuelsWon"] / eligible_df["minutesPlayed"]
) * 90

In [58]:
eligible_df[
    [
        "player_name",
        "position",
        "groundDuelsWon",
        "minutesPlayed",
        "groundDuelsWon_per90"
    ]
].head(10)

,player_name,position,groundDuelsWon,minutesPlayed,groundDuelsWon_per90
0,Declan Rice,M,77,2585,2.680851
1,Gabriel Magalhães,D,60,2166,2.493075
2,Bukayo Saka,F,136,2001,6.116942
3,Martín Zubimendi,M,68,2604,2.350230
5,Martin Ødegaard,M,33,1078,2.755102
6,Eberechi Eze,F,69,1504,4.128989
7,William Saliba,D,48,2075,2.081928
8,David Raya,G,11,2790,0.354839
9,Piero Hincapié,D,55,1504,3.291223
11,Jurriën Timber,D,107,2457,3.919414


In [59]:
eligible_df[
    [
        "player_name",
        "position",
        "groundDuelsWon_per90",
        "groundDuelsWonPercentage"
    ]
].head(10)

,player_name,position,groundDuelsWon_per90,groundDuelsWonPercentage
0,Declan Rice,M,2.680851,56.204380
1,Gabriel Magalhães,D,2.493075,75.949367
2,Bukayo Saka,F,6.116942,54.618474
3,Martín Zubimendi,M,2.350230,54.400000
5,Martin Ødegaard,M,2.755102,46.478873
6,Eberechi Eze,F,4.128989,57.500000
7,William Saliba,D,2.081928,60.000000
8,David Raya,G,0.354839,100.000000
9,Piero Hincapié,D,3.291223,61.797753
11,Jurriën Timber,D,3.919414,54.314721


In [60]:
eligible_df["goals_per90"].rank(pct=True) * 100

0      68.106312
1      65.448505
2      85.714286
3      74.750831
5      52.823920
         ...    
508    51.827243
509    72.425249
511    16.445183
512    80.398671
515    16.445183
Name: goals_per90, Length: 301, dtype: float64

In [61]:
eligible_df["goals_per90_pct"] = (
    eligible_df.groupby("position")["goals_per90"]
    .rank(pct=True) * 100
)

In [62]:
eligible_df[
    [
        "player_name",
        "position",
        "goals_per90",
        "goals_per90_pct"
    ]
].query(
    "player_name in ['Bukayo Saka', 'Declan Rice', 'William Saliba', 'David Raya']"
)

,player_name,position,goals_per90,goals_per90_pct
0,Declan Rice,M,0.139265,60.162602
2,Bukayo Saka,F,0.269865,42.553191
7,William Saliba,D,0.043373,62.037037
8,David Raya,G,0.000000,52.173913


In [63]:
eligible_df.query("position == 'F'")[
    ["player_name", "goals_per90", "goals_per90_pct"]
].sort_values("goals_per90", ascending=False).head(10)

,player_name,goals_per90,goals_per90_pct
26,Erling Haaland,0.818859,100.000000
310,Eli Junior Kroupi,0.693493,97.872340
154,Igor Thiago,0.644310,95.744681
65,Benjamin Šeško,0.608108,93.617021
233,Danny Welbeck,0.591457,91.489362
468,Zian Flemming,0.589198,89.361702
112,Hugo Ekitiké,0.546961,87.234043
132,João Pedro,0.542402,85.106383
409,Richarlison,0.538564,82.978723
21,Viktor Gyökeres,0.519140,80.851064


In [64]:
eligible_df["assists_per90_pct"] = (
    eligible_df.groupby("position")["assists_per90"]
    .rank(pct=True) * 100
)

In [65]:
eligible_df[
    [
        "player_name",
        "position",
        "accuratePasses",
        "accuratePassesPercentage",
        "accurateFinalThirdPasses",
        "accurateLongBalls",
        "accurateLongBallsPercentage"
    ]
].head(10)

,player_name,position,accuratePasses,accuratePassesPercentage,accurateFinalThirdPasses,accurateLongBalls,accurateLongBallsPercentage
0,Declan Rice,M,1578,87.813022,459,72,48.648649
1,Gabriel Magalhães,D,1259,88.724454,180,63,46.323529
2,Bukayo Saka,F,524,76.720351,300,14,36.842105
3,Martín Zubimendi,M,1494,88.193625,369,48,49.484536
5,Martin Ødegaard,M,537,85.103011,228,32,56.140351
6,Eberechi Eze,F,437,84.200385,184,13,39.393939
7,William Saliba,D,1623,92.689891,221,44,48.351648
8,David Raya,G,604,66.519824,72,163,35.281385
9,Piero Hincapié,D,662,89.218329,146,20,35.714286
11,Jurriën Timber,D,925,85.253456,344,20,38.461538


In [66]:
eligible_df["accuratePasses_per90"] = (
    eligible_df["accuratePasses"] / eligible_df["minutesPlayed"]
) * 90

In [67]:
eligible_df[
    [
        "player_name",
        "position",
        "accuratePasses",
        "minutesPlayed",
        "accuratePasses_per90"
    ]
].head(10)

,player_name,position,accuratePasses,minutesPlayed,accuratePasses_per90
0,Declan Rice,M,1578,2585,54.940039
1,Gabriel Magalhães,D,1259,2166,52.313019
2,Bukayo Saka,F,524,2001,23.568216
3,Martín Zubimendi,M,1494,2604,51.635945
5,Martin Ødegaard,M,537,1078,44.833024
6,Eberechi Eze,F,437,1504,26.150266
7,William Saliba,D,1623,2075,70.395181
8,David Raya,G,604,2790,19.483871
9,Piero Hincapié,D,662,1504,39.614362
11,Jurriën Timber,D,925,2457,33.882784


In [68]:
eligible_df["accuratePasses_per90_pct"] = (
    eligible_df.groupby("position")["accuratePasses_per90"]
    .rank(pct=True) * 100
)

In [69]:
eligible_df[
    [
        "player_name",
        "position",
        "accuratePasses_per90",
        "accuratePasses_per90_pct"
    ]
].head(10)

,player_name,position,accuratePasses_per90,accuratePasses_per90_pct
0,Declan Rice,M,54.940039,94.308943
1,Gabriel Magalhães,D,52.313019,82.407407
2,Bukayo Saka,F,23.568216,82.978723
3,Martín Zubimendi,M,51.635945,91.869919
5,Martin Ødegaard,M,44.833024,78.861789
6,Eberechi Eze,F,26.150266,95.744681
7,William Saliba,D,70.395181,99.074074
8,David Raya,G,19.483871,43.478261
9,Piero Hincapié,D,39.614362,55.555556
11,Jurriën Timber,D,33.882784,30.555556


In [70]:
eligible_df["accurateFinalThirdPasses_per90"] = (
    eligible_df["accurateFinalThirdPasses"] / eligible_df["minutesPlayed"]
) * 90

In [71]:
eligible_df[
    [
        "player_name",
        "position",
        "accurateFinalThirdPasses",
        "minutesPlayed",
        "accurateFinalThirdPasses_per90"
    ]
].head(10)

,player_name,position,accurateFinalThirdPasses,minutesPlayed,accurateFinalThirdPasses_per90
0,Declan Rice,M,459,2585,15.980658
1,Gabriel Magalhães,D,180,2166,7.479224
2,Bukayo Saka,F,300,2001,13.493253
3,Martín Zubimendi,M,369,2604,12.753456
5,Martin Ødegaard,M,228,1078,19.035250
6,Eberechi Eze,F,184,1504,11.010638
7,William Saliba,D,221,2075,9.585542
8,David Raya,G,72,2790,2.322581
9,Piero Hincapié,D,146,1504,8.736702
11,Jurriën Timber,D,344,2457,12.600733


In [72]:
eligible_df.groupby("position")["accurateFinalThirdPasses_per90"].describe()

,count,mean,std,min,25%,50%,75%,max
position,,,,,,,,
D,108.0,7.082156,3.390231,1.937086,4.231850,6.160310,9.116042,18.262020
F,47.0,6.677810,3.344668,2.432432,4.162880,5.611336,9.886641,14.390459
G,23.0,1.591004,0.836196,0.444444,1.041129,1.437500,1.921053,4.032258
M,123.0,10.740275,4.112741,4.692443,7.347156,9.879699,12.943587,24.276316


In [73]:
eligible_df.groupby("position")["accuratePassesPercentage"].describe()

,count,mean,std,min,25%,50%,75%,max
position,,,,,,,,
D,108.0,84.450427,4.845219,73.635153,81.389156,85.350266,87.982499,95.133953
F,47.0,72.149291,8.102198,57.009346,65.506196,72.839506,76.806877,88.480392
G,23.0,63.719995,8.756141,46.666667,56.967679,62.972973,67.489194,79.617834
M,123.0,83.003542,5.016642,66.737288,79.845438,83.556405,86.490665,92.249527


In [74]:
eligible_df[
    [
        "player_name",
        "position",
        "assists",
        "bigChancesCreated",
        "keyPasses"
    ]
].head(10)

,player_name,position,assists,bigChancesCreated,keyPasses
0,Declan Rice,M,5,14,59
1,Gabriel Magalhães,D,4,5,6
2,Bukayo Saka,F,3,9,54
3,Martín Zubimendi,M,1,1,16
5,Martin Ødegaard,M,5,5,31
6,Eberechi Eze,F,2,4,13
7,William Saliba,D,0,0,7
8,David Raya,G,0,0,3
9,Piero Hincapié,D,2,2,12
11,Jurriën Timber,D,5,4,27


In [75]:
eligible_df.groupby("position")[
    ["assists", "bigChancesCreated", "keyPasses"]
].mean()

,assists,bigChancesCreated,keyPasses
position,,,
D,1.064815,1.944444,11.055556
F,1.574468,3.063830,17.489362
G,0.086957,0.173913,0.956522
M,2.365854,4.178862,23.479675


In [76]:
eligible_df["keyPasses_per90"] = (
    eligible_df["keyPasses"] / eligible_df["minutesPlayed"]
) * 90

In [77]:
eligible_df[
    [
        "player_name",
        "position",
        "keyPasses",
        "minutesPlayed",
        "keyPasses_per90"
    ]
].head(10)

,player_name,position,keyPasses,minutesPlayed,keyPasses_per90
0,Declan Rice,M,59,2585,2.054159
1,Gabriel Magalhães,D,6,2166,0.249307
2,Bukayo Saka,F,54,2001,2.428786
3,Martín Zubimendi,M,16,2604,0.552995
5,Martin Ødegaard,M,31,1078,2.588126
6,Eberechi Eze,F,13,1504,0.777926
7,William Saliba,D,7,2075,0.303614
8,David Raya,G,3,2790,0.096774
9,Piero Hincapié,D,12,1504,0.718085
11,Jurriën Timber,D,27,2457,0.989011


In [78]:
eligible_df.groupby("position")["keyPasses_per90"].describe()

,count,mean,std,min,25%,50%,75%,max
position,,,,,,,,
D,108.0,0.543059,0.388681,0.000000,0.259795,0.437714,0.723932,1.663244
F,47.0,1.006716,0.496753,0.000000,0.670151,0.966999,1.204975,2.428786
G,23.0,0.034843,0.050239,0.000000,0.000000,0.000000,0.047091,0.178218
M,123.0,1.221836,0.629988,0.163339,0.779859,1.144254,1.622891,3.729996


In [80]:
eligible_df["keyPasses_per90_pct"] = (
    eligible_df.groupby("position")["keyPasses_per90"]
    .rank(pct=True) * 100
)

In [81]:
eligible_df[
    [
        "player_name",
        "position",
        "keyPasses_per90",
        "keyPasses_per90_pct"
    ]
].head(10)

,player_name,position,keyPasses_per90,keyPasses_per90_pct
0,Declan Rice,M,2.054159,91.056911
1,Gabriel Magalhães,D,0.249307,23.148148
2,Bukayo Saka,F,2.428786,100.000000
3,Martín Zubimendi,M,0.552995,14.634146
5,Martin Ødegaard,M,2.588126,97.560976
6,Eberechi Eze,F,0.777926,38.297872
7,William Saliba,D,0.303614,33.333333
8,David Raya,G,0.096774,86.956522
9,Piero Hincapié,D,0.718085,73.148148
11,Jurriën Timber,D,0.989011,85.185185


In [82]:
eligible_df["bigChancesCreated_per90"] = (
    eligible_df["bigChancesCreated"] / eligible_df["minutesPlayed"]
) * 90

In [83]:
eligible_df[
    [
        "player_name",
        "position",
        "bigChancesCreated",
        "minutesPlayed",
        "bigChancesCreated_per90"
    ]
].head(10)

,player_name,position,bigChancesCreated,minutesPlayed,bigChancesCreated_per90
0,Declan Rice,M,14,2585,0.487427
1,Gabriel Magalhães,D,5,2166,0.207756
2,Bukayo Saka,F,9,2001,0.404798
3,Martín Zubimendi,M,1,2604,0.034562
5,Martin Ødegaard,M,5,1078,0.417440
6,Eberechi Eze,F,4,1504,0.239362
7,William Saliba,D,0,2075,0.000000
8,David Raya,G,0,2790,0.000000
9,Piero Hincapié,D,2,1504,0.119681
11,Jurriën Timber,D,4,2457,0.146520


In [84]:
eligible_df.groupby("position")["bigChancesCreated_per90"].describe()

,count,mean,std,min,25%,50%,75%,max
position,,,,,,,,
D,108.0,0.095077,0.092568,0.0,0.033065,0.071268,0.150184,0.456467
F,47.0,0.169123,0.140524,0.0,0.083271,0.141584,0.191848,0.645418
G,23.0,0.006293,0.017672,0.0,0.000000,0.000000,0.000000,0.071287
M,123.0,0.216593,0.180562,0.0,0.085097,0.185185,0.329273,1.025836


In [85]:
eligible_df["bigChancesCreated_per90_pct"] = (
    eligible_df.groupby("position")["bigChancesCreated_per90"]
    .rank(pct=True) * 100
)

In [86]:
eligible_df[
    ["player_name", "position", "bigChancesCreated_per90",
     "bigChancesCreated_per90_pct"]
].head(10)

,player_name,position,bigChancesCreated_per90,bigChancesCreated_per90_pct
0,Declan Rice,M,0.487427,93.495935
1,Gabriel Magalhães,D,0.207756,88.888889
2,Bukayo Saka,F,0.404798,91.489362
3,Martín Zubimendi,M,0.034562,12.195122
5,Martin Ødegaard,M,0.417440,88.617886
6,Eberechi Eze,F,0.239362,85.106383
7,William Saliba,D,0.000000,12.500000
8,David Raya,G,0.000000,45.652174
9,Piero Hincapié,D,0.119681,67.592593
11,Jurriën Timber,D,0.146520,72.222222


In [87]:
eligible_df[
    ["bigChancesCreated_per90", "keyPasses_per90"]
].corr()

,bigChancesCreated_per90,keyPasses_per90
bigChancesCreated_per90,1.000000,0.771451
keyPasses_per90,0.771451,1.000000


In [88]:
eligible_df["assists_per90"] = (
    eligible_df["assists"] / eligible_df["minutesPlayed"] * 90
)

In [89]:
eligible_df[
    ["player_name", "position", "assists", "minutesPlayed", "assists_per90"]
].head(10)

,player_name,position,assists,minutesPlayed,assists_per90
0,Declan Rice,M,5,2585,0.174081
1,Gabriel Magalhães,D,4,2166,0.166205
2,Bukayo Saka,F,3,2001,0.134933
3,Martín Zubimendi,M,1,2604,0.034562
5,Martin Ødegaard,M,5,1078,0.417440
6,Eberechi Eze,F,2,1504,0.119681
7,William Saliba,D,0,2075,0.000000
8,David Raya,G,0,2790,0.000000
9,Piero Hincapié,D,2,1504,0.119681
11,Jurriën Timber,D,5,2457,0.183150


In [90]:
eligible_df.groupby("position")["assists_per90"].describe()

,count,mean,std,min,25%,50%,75%,max
position,,,,,,,,
D,108.0,0.050377,0.061327,0.0,0.000000,0.036715,0.074090,0.239521
F,47.0,0.087708,0.077601,0.0,0.036478,0.072522,0.135203,0.318021
G,23.0,0.003724,0.012526,0.0,0.000000,0.000000,0.000000,0.050000
M,123.0,0.124971,0.102501,0.0,0.056356,0.107100,0.169173,0.590890


In [91]:
eligible_df["assists_per90_pct"] = (
    eligible_df.groupby("position")["assists_per90"]
    .rank(pct=True) * 100
)

In [92]:
eligible_df[
    ["player_name", "position", "assists_per90", "assists_per90_pct"]
].head(10)

,player_name,position,assists_per90,assists_per90_pct
0,Declan Rice,M,0.174081,76.422764
1,Gabriel Magalhães,D,0.166205,92.592593
2,Bukayo Saka,F,0.134933,74.468085
3,Martín Zubimendi,M,0.034562,15.447154
5,Martin Ødegaard,M,0.417440,98.373984
6,Eberechi Eze,F,0.119681,72.340426
7,William Saliba,D,0.000000,23.148148
8,David Raya,G,0.000000,47.826087
9,Piero Hincapié,D,0.119681,84.259259
11,Jurriën Timber,D,0.183150,96.296296


In [93]:
eligible_df[
    ["bigChancesCreated_per90", "keyPasses_per90", "assists_per90"]
].corr()

,bigChancesCreated_per90,keyPasses_per90,assists_per90
bigChancesCreated_per90,1.000000,0.771451,0.715788
keyPasses_per90,0.771451,1.000000,0.656295
assists_per90,0.715788,0.656295,1.000000


In [94]:
eligible_df["chance_creation_score"] = (
    eligible_df["bigChancesCreated_per90_pct"] * 0.50
    + eligible_df["assists_per90_pct"] * 0.30
    + eligible_df["keyPasses_per90_pct"] * 0.20
)

In [95]:
eligible_df[
    ["player_name", "position",
     "bigChancesCreated_per90_pct",
     "assists_per90_pct",
     "keyPasses_per90_pct",
     "chance_creation_score"]
].head(10)

,player_name,position,bigChancesCreated_per90_pct,assists_per90_pct,keyPasses_per90_pct,chance_creation_score
0,Declan Rice,M,93.495935,76.422764,91.056911,87.886179
1,Gabriel Magalhães,D,88.888889,92.592593,23.148148,76.851852
2,Bukayo Saka,F,91.489362,74.468085,100.000000,88.085106
3,Martín Zubimendi,M,12.195122,15.447154,14.634146,13.658537
5,Martin Ødegaard,M,88.617886,98.373984,97.560976,93.333333
6,Eberechi Eze,F,85.106383,72.340426,38.297872,71.914894
7,William Saliba,D,12.500000,23.148148,33.333333,19.861111
8,David Raya,G,45.652174,47.826087,86.956522,54.565217
9,Piero Hincapié,D,67.592593,84.259259,73.148148,73.703704
11,Jurriën Timber,D,72.222222,96.296296,85.185185,82.037037


In [96]:
eligible_df[
    ["player_name", "position", "goals_per90", "goals_per90_pct"]
].head(10)

,player_name,position,goals_per90,goals_per90_pct
0,Declan Rice,M,0.139265,60.162602
1,Gabriel Magalhães,D,0.124654,91.666667
2,Bukayo Saka,F,0.269865,42.553191
3,Martín Zubimendi,M,0.172811,69.918699
5,Martin Ødegaard,M,0.083488,39.024390
6,Eberechi Eze,F,0.359043,61.702128
7,William Saliba,D,0.043373,62.037037
8,David Raya,G,0.000000,52.173913
9,Piero Hincapié,D,0.059840,69.444444
11,Jurriën Timber,D,0.109890,89.814815


In [97]:
eligible_df["totalShots_per90"] = (
    eligible_df["totalShots"] / eligible_df["minutesPlayed"] * 90
)

In [98]:
eligible_df[
    ["player_name", "position", "totalShots", "minutesPlayed", "totalShots_per90"]
].head(10)

,player_name,position,totalShots,minutesPlayed,totalShots_per90
0,Declan Rice,M,36,2585,1.253385
1,Gabriel Magalhães,D,18,2166,0.747922
2,Bukayo Saka,F,63,2001,2.833583
3,Martín Zubimendi,M,26,2604,0.898618
5,Martin Ødegaard,M,20,1078,1.669759
6,Eberechi Eze,F,49,1504,2.932181
7,William Saliba,D,5,2075,0.216867
8,David Raya,G,0,2790,0.000000
9,Piero Hincapié,D,2,1504,0.119681
11,Jurriën Timber,D,25,2457,0.915751


In [99]:
eligible_df["totalShots_per90"] = (
    eligible_df["totalShots"] / eligible_df["minutesPlayed"] * 90
)

In [100]:
eligible_df[
    ["player_name", "position", "totalShots", "minutesPlayed", "totalShots_per90"]
].head(10)

,player_name,position,totalShots,minutesPlayed,totalShots_per90
0,Declan Rice,M,36,2585,1.253385
1,Gabriel Magalhães,D,18,2166,0.747922
2,Bukayo Saka,F,63,2001,2.833583
3,Martín Zubimendi,M,26,2604,0.898618
5,Martin Ødegaard,M,20,1078,1.669759
6,Eberechi Eze,F,49,1504,2.932181
7,William Saliba,D,5,2075,0.216867
8,David Raya,G,0,2790,0.000000
9,Piero Hincapié,D,2,1504,0.119681
11,Jurriën Timber,D,25,2457,0.915751


In [101]:
eligible_df.groupby("position")["totalShots_per90"].describe()

,count,mean,std,min,25%,50%,75%,max
position,,,,,,,,
D,108.0,0.610414,0.327501,0.000000,0.381297,0.553695,0.802494,1.731959
F,47.0,2.281609,0.648376,1.093117,1.816857,2.366274,2.724685,3.796526
G,23.0,0.001989,0.009541,0.000000,0.000000,0.000000,0.000000,0.045755
M,123.0,1.379389,0.671994,0.105448,0.901343,1.284567,1.838209,3.633803


In [102]:
eligible_df["totalShots_per90_pct"] = (
    eligible_df.groupby("position")["totalShots_per90"]
    .rank(pct=True) * 100
)

In [103]:
eligible_df[
    ["player_name", "position",
     "totalShots_per90",
     "totalShots_per90_pct"]
].head(10)

,player_name,position,totalShots_per90,totalShots_per90_pct
0,Declan Rice,M,1.253385,48.780488
1,Gabriel Magalhães,D,0.747922,69.444444
2,Bukayo Saka,F,2.833583,85.106383
3,Martín Zubimendi,M,0.898618,25.203252
5,Martin Ødegaard,M,1.669759,69.918699
6,Eberechi Eze,F,2.932181,89.361702
7,William Saliba,D,0.216867,10.185185
8,David Raya,G,0.000000,50.000000
9,Piero Hincapié,D,0.119681,5.555556
11,Jurriën Timber,D,0.915751,84.259259


In [104]:
eligible_df[
    ["goals_per90", "totalShots_per90"]
].corr()

,goals_per90,totalShots_per90
goals_per90,1.00000,0.77861
totalShots_per90,0.77861,1.00000


In [105]:
for i, col in enumerate(eligible_df.columns):
    print(i, col)

0 player_name
1 team_name
2 position
3 accurateCrosses
4 accurateCrossesPercentage
5 accurateFinalThirdPasses
6 accurateLongBalls
7 accurateLongBallsPercentage
8 accurateOppositionHalfPasses
9 accurateOwnHalfPasses
10 accuratePasses
11 accuratePassesPercentage
12 aerialDuelsWon
13 aerialDuelsWonPercentage
14 aerialLost
15 appearances
16 assists
17 ballRecovery
18 bigChancesCreated
19 bigChancesMissed
20 blockedShots
21 cleanSheet
22 clearances
23 dispossessed
24 dribbledPast
25 duelLost
26 errorLeadToGoal
27 errorLeadToShot
28 expectedAssists
29 expectedGoals
30 fouls
31 freeKickGoal
32 goalConversionPercentage
33 goals
34 goalsAssistsSum
35 goalsConceded
36 goalsConcededInsideTheBox
37 goalsConcededOutsideTheBox
38 goalsFromInsideTheBox
39 goalsFromOutsideTheBox
40 goalsPrevented
41 groundDuelsWon
42 groundDuelsWonPercentage
43 headedGoals
44 highClaims
45 hitWoodwork
46 inaccuratePasses
47 interceptions
48 keyPasses
49 leftFootGoals
50 matchesStarted
51 minutesPlayed
52 offsides
53 o

In [106]:
eligible_df["xG_per90"] = (
    eligible_df["expectedGoals"] / eligible_df["minutesPlayed"] * 90
)

eligible_df["successfulDribbles_per90"] = (
    eligible_df["successfulDribbles"] / eligible_df["minutesPlayed"] * 90
)

eligible_df["possessionLost_per90"] = (
    eligible_df["possessionLost"] / eligible_df["minutesPlayed"] * 90
)

In [107]:
eligible_df[
    [
        "player_name",
        "position",
        "xG_per90",
        "successfulDribbles_per90",
        "possessionLost_per90"
    ]
].head(10)

,player_name,position,xG_per90,successfulDribbles_per90,possessionLost_per90
0,Declan Rice,M,0.100034,0.417795,14.135397
1,Gabriel Magalhães,D,0.078108,0.041551,7.520776
2,Bukayo Saka,F,0.314577,2.203898,16.281859
3,Martín Zubimendi,M,0.096124,0.207373,8.917051
5,Martin Ødegaard,M,0.084390,0.751391,12.356215
6,Eberechi Eze,F,0.224443,1.376330,8.916223
7,William Saliba,D,0.037193,0.173494,6.766265
8,David Raya,G,NaN,0.032258,10.064516
9,Piero Hincapié,D,0.019664,0.299202,8.736702
11,Jurriën Timber,D,0.172267,0.476190,11.062271


In [108]:
forward_df = eligible_df[eligible_df["position"] == "F"].copy()

forward_df["xG_per90_pct"] = (
    forward_df["xG_per90"].rank(pct=True) * 100
)

forward_df["successfulDribbles_per90_pct"] = (
    forward_df["successfulDribbles_per90"].rank(pct=True) * 100
)

forward_df["possessionLost_per90_pct"] = (
    forward_df["possessionLost_per90"].rank(pct=True) * 100
)

In [109]:
forward_df[
    [
        "player_name",
        "xG_per90",
        "xG_per90_pct",
        "successfulDribbles_per90",
        "successfulDribbles_per90_pct",
        "possessionLost_per90",
        "possessionLost_per90_pct"
    ]
].head(10)

,player_name,xG_per90,xG_per90_pct,successfulDribbles_per90,successfulDribbles_per90_pct,possessionLost_per90,possessionLost_per90_pct
2,Bukayo Saka,0.314577,51.063830,2.203898,95.744681,16.281859,89.361702
6,Eberechi Eze,0.224443,31.914894,1.376330,82.978723,8.916223,12.765957
12,Leandro Trossard,0.257566,42.553191,0.908012,65.957447,16.130564,87.234043
21,Viktor Gyökeres,0.441260,70.212766,0.283167,10.638298,11.043524,34.042553
26,Erling Haaland,0.781991,100.000000,0.483871,27.659574,6.774194,2.127660
53,Matheus Cunha,0.257109,40.425532,1.648193,85.106383,15.440964,85.106383
54,Amad Diallo,0.219775,29.787234,1.688379,87.234043,10.739969,27.659574
65,Benjamin Šeško,0.484318,78.723404,0.405405,14.893617,10.675676,25.531915
90,Ollie Watkins,0.403344,65.957447,0.819957,55.319149,9.214751,14.893617
112,Hugo Ekitiké,0.492280,82.978723,1.889503,91.489362,10.740331,29.787234


In [110]:
forward_df["bigChancesMissed_per90"] = (
    forward_df["bigChancesMissed"] / forward_df["minutesPlayed"] * 90
)

In [112]:
forward_df["bigChancesMissed_per90_pct"] = (
    forward_df["bigChancesMissed_per90"].rank(pct=True) * 100
)